tools/wake/timeline.cpp

```diff
@@ -0,0 +1,242 @@
+        if (!(job_pair.second.top_sort_visited)) {
+            dfs_top_sort(job_pair.first, job_map, top_sorted_jobs);
+        }
+    }
+}
+
+long fill_one_dependency(std::pair<int, long> access, long dependency, std::map<long, JobNode> &job_map) {
+    if (access.first == 2) {
+        return access.second;
+    }
+    if (job_map.find(access.second) != job_map.end()) {
+        job_map.at(access.second).dependencies.emplace_back(dependency);
+    }
+    return dependency;
+}
+
+void fill_all_dependencies(const std::vector<std::pair<int, long>> &accesses, std::map<long, JobNode> &job_map) {
+    long dependency = -1;
+    for (auto access: accesses) {
+        dependency = fill_one_dependency(access, dependency, job_map);
+    }
+}
+
+void assign_parents(std::map<long, JobNode> &job_map, std::vector<long> &top_sorted_jobs) {
+    for (long job_id: top_sorted_jobs) {
+        JobNode &job_node = job_map.at(job_id);
+        for (long dependency_id: job_node.dependencies) {
+            JobNode dependency = job_map.at(dependency_id);
+            if (dependency.hypothetical_endtime >= job_node.hypothetical_starttime) {
+                job_node.hypothetical_starttime = dependency.hypothetical_endtime;
+                job_node.parent_id = dependency_id;
+            }
+        }
+
+        job_node.hypothetical_endtime = job_node.hypothetical_starttime + (job_node.endtime - job_node.starttime);
+    }
+}
+
+void find_critical_path(std::map<long, JobNode> &job_map, std::vector<long> &critical_path) {
+    long latest_job_id = -1;
+    for (const auto &job_pair: job_map) {
+        if (latest_job_id == -1 ||
+            job_map.at(latest_job_id).hypothetical_endtime < job_pair.second.hypothetical_endtime) {
+            latest_job_id = job_pair.second.job.job;
+        }
+    }
+    job_map.at(latest_job_id).is_on_critical_path = true;
+    critical_path.emplace_back(job_map.at(latest_job_id).job.job);
+    while (job_map.at(latest_job_id).parent_id != -1) {
+        latest_job_id = job_map.at(latest_job_id).parent_id;
+        job_map.at(latest_job_id).is_on_critical_path = true;
+        critical_path.emplace_back(job_map.at(latest_job_id).job.job);
+    }
+}
+
+void dfs_transitive_reduction(long job_id, std::map<long, JobNode> &job_map, std::vector<int> &ancestors) {
+    JobNode &job = job_map.at(job_id);
+    job_map.at(job_id).top_sort_visited = true;
+    ancestors.emplace_back(job_id);
+
+    for (long child_id: job.dependencies) {
+        JobNode &child = job_map.at(child_id);
+        for (long grandchild_id: child.dependencies) {
+            for (int ancestor_id: ancestors) {
+                JobNode ancestor = job_map.at(ancestor_id);
+                if (std::find(ancestor.dependencies.begin(), ancestor.dependencies.end(), grandchild_id) !=
+                    ancestor.dependencies.end()) {
+                    std::remove(ancestor.dependencies.begin(), ancestor.dependencies.end(), grandchild_id);
+                }
```

Comment:
Maybe this is what you wanted but std::remove doesn't actually remove anything from ancestor, it just moves elements over the top of your item leaving the last set of items in whatever state std::move leaves them in. In this case because std::move on int is just a copy you're just adding duplicate deps. I think in this case its just setting the ancestor ids to zero. To actually update ancestors you need to do this

```
auto new_end = std::remove(...);
ancestors.dependencies.erase(new_end, ancestors.dependencies.end());
```

That's a good way to remove many elements but is not really ideal here since we're only removing a single element. Something like this is better:

```
auto iter = std::find(...);
if (iter != ancestor.dependencies.end()) {
    ancestor.dependencies.erase(iter);
}
```

Not needed but a possible clean up:
An std::set or std::unordered_set might be a better data structure here than an std::vector for the deps. They're still iterable, you can add things to them with emplace_back, you can check for membership in O(lg n) or expected O(1) respectively, and you can remove from them just as quickly. I think that covers all the operations you. But efficiency is less than concern really, I just think the code might be a bit cleaner that way.

Reply:
Thank you! Done